src/net.c

```diff
@@ -115,6 +115,28 @@ timeout_connect(int s, const struct sockaddr *name, socklen_t namelen,
 	return (ret);
 }
 
+#ifdef __CYGWIN__
+
+const unsigned int MAX_GETADDRINFO_RETRIES = 10;
+/* getaddrinfo may not be completely reliable on Windows hosts. Try more than once. */
+int iperf_getaddrinfo(const char *restrict node,
+                       const char *restrict service,
+                       const struct addrinfo *restrict hints,
+                       struct addrinfo **restrict res)
+{
+    int retvalue = -1;
+    int i = 0;
+    while (i++ < MAX_GETADDRINFO_RETRIES) {
+        retvalue = getaddrinfo(node, service, hints, res);
+        if (0 == retvalue) break;
+        sleep(1); /* yield thread */
+        fprintf(stderr, "getadddrinfo failed. If iperf still ran successfully, this may be a Windows bug. Please see https://github.com/esnet/iperf/issues/1314 . \n");
```

Comment:
Error should be either printer only if debug option was set or using `iperf_err()` (to properly handle the JSON output case).  **Suggest to leave the decision to the iperf3 maintainers** (before merging the change).

In both cases, `test` should be passed as a parameter to the function and `iperf.h` should be included.  In case `iperf_err` is used, also `iperf_api.h` should be included

If printed as debug message, suggest to use the newly added debug level option:
```
if (test->debug_level >= DEBUG_LEVEL_ERROR))
    fprintf(...);
```

Reply:
David, thanks for the feedback. Obviously, I wrote my PR before the option was added. I would prefer for this message to be printed unconditionally if getaddrinfo fails intermittently, regardless of the new debug option. The rationale is that with the loop in place, a fatal error would not be seen anymore. And without being aware of the intermittent failure behind the scene, nobody would think to add -d. The intermittent failure seems rare enough that I wanted anyone with a affected system to see the message. Requiring -d would significantly decrease the odds anyone would ever know about the problem. The wording of the message could be changed, though.

I checked the source to see where debug_level was being set. The iperf_test structure that contains debug_level isn't available in the scope of most of the calls to getaddrinfo . A pointer to it would need to be propagated across the stack. getaddrinfo itself of course, being a low-level call, doesn't take this object as an argument. I am reluctant to change the prototype for getaddrinfo . I could make it a wrapper on all platforms, instead of having the platform #ifdef, though.

iperf_test appears to contain a global configuration state. And there is only ever a single instance, created in main() with a call to iperf_new_test. As long as that remains the case, it strikes me as a bit of busy work to pass this object around the stack down everywhere, not just to iperf_getaddrinfo.  I don't want to address that in this PR.

Reply:
> I would prefer for this message to be printed unconditionally if getaddrinfo fails intermittently, regardless of the new debug option.

The new debug option is just the ability to set a debug level.  Using `if (test->debug)` always existed, but it still needs the use of `iperf_test`.

> The rationale is that with the loop in place, a fatal error would not be seen anymore. And without being aware of the intermittent failure behind the scene, nobody would think to add -d.

I agree.  In this case still `iperf_err()` should be used to log the error, to properly handle JSON output (in case `-J` was set), which also needs the use of  `iperf_test`.

> A pointer to it would need to be propagated across the stack. getaddrinfo itself of course, being a low-level call, doesn't take this object as an argument. I am reluctant to change the prototype for getaddrinfo . .....

I agree that adding `iperf_test` to `getaddrinfo` and to net.c in general is a problem.  I therefore suggest to leave the decision to the iperf3 maintainers whether to propagate `iperf_test` down to net.c, not log the error in `getaddrinfo`, add a wrapper or any of the other approaches you suggested.